leopard.nim

```diff
@@ -0,0 +1,329 @@
+    code*: LeopardResult
+
+  ParityData* = Data
+
+  ReedSolomonCode* = tuple[codeword, data, parity: uint] # symbol counts
+
+# workaround for https://github.com/nim-lang/Nim/issues/19619
+# necessary for use of nim-leopard in nimbus-build-system projects because nbs
+# ships libbacktrace by default
+proc `$`*(err: LeopardError): string {.nosideeffect.} =
+  $err
+
+# https://github.com/catid/leopard/issues/12
+# https://www.cs.cmu.edu/~guyb/realworld/reedsolomon/reed_solomon_codes.html
+#
+# RS(255,239)
+# ---------------------------------
+# codeword symbols = 255
+# data symbols     = 239
+# parity symbols   = 255 - 239 = 16
+
+proc RS*(codeword, data: Positive): ReedSolomonCode =
+  var
+    parity = codeword - data
+
+  if parity < 0: parity = 0
+  (codeword: codeword.uint, data: data.uint, parity: parity.uint)
+
+func isValid*(code: ReedSolomonCode): bool =
+  not ((code.codeword - code.data != code.parity) or
+       (code.parity > code.data) or (code.codeword < MinSymbols + 1) or
+       (code.data < MinSymbols) or (code.parity < MinSymbols) or
+       (code.codeword > MaxTotalSymbols))
+
+# alloc/freeAligned and helpers adapted from mratsim/weave:
+# https://github.com/mratsim/weave/blob/master/weave/memory/allocs.nim
+
+func isPowerOfTwo(n: int): bool {.inline.} =
+  (n and (n - 1)) == 0
+
+func roundNextMultipleOf(x: Natural, n: Natural): int {.inline.} =
+  (x + n - 1) and not (n - 1)
+
+when defined(windows):
+  proc aligned_alloc_windows(size, alignment: csize_t): pointer
+    {.header: "<malloc.h>", importc: "_aligned_malloc", sideeffect.}
+
+  proc aligned_free_windows(p: pointer)
+    {.header: "<malloc.h>", importc: "_aligned_free", sideeffect.}
+
+  proc freeAligned*(p: pointer) =
+    if not p.isNil:
+      aligned_free_windows(p)
+
+elif defined(osx):
+  proc posix_memalign(mem: var pointer, alignment, size: csize_t)
+    {.header: "<stdlib.h>", importc, sideeffect.}
+
+  proc aligned_alloc(alignment, size: csize_t): pointer {.inline.} =
+    posix_memalign(result, alignment, size)
+
+else:
+  proc aligned_alloc(alignment, size: csize_t): pointer
+    {.header: "<stdlib.h>", importc, sideeffect.}
+
+when not defined(windows):
+  proc c_free(p: pointer) {.header: "<stdlib.h>", importc: "free".}
+
+  proc freeAligned*(p: pointer) {.inline.} =
+    if not p.isNil:
+      c_free(p)
+
+proc allocAligned*(size: int, alignment: static Natural): pointer {.inline.} =
+  ## aligned_alloc requires allocating in multiple of the alignment.
+  static:
+    assert alignment.isPowerOfTwo()
+
+  let
+    requiredMem = size.roundNextMultipleOf(alignment)
+
+  when defined(windows):
+    aligned_alloc_windows(csize_t requiredMem, csize_t alignment)
+  else:
+    aligned_alloc(csize_t alignment, csize_t requiredMem)
+
+proc leoInit*() =
+  if wrapper.leoInit() != 0:
+    raise (ref LeopardDefect)(msg: "Leopard-RS failed to initialize")
+
+proc encode*(code: ReedSolomonCode, data: Data):
+    Result[ParityData, LeopardError] =
```

Comment:
We're incurring into potentially a very large copy overhead by returning `ParityData` (which could be of the order of GBs). In this case, the best would be to allow the caller to pre-allocate and pass the buffer either as a `var` (preferable) or a straight up pointer.

Reply:
Agreed. I was/am working today on a refactor to allows for `var` params in `proc encode` and `proc decode`. I'll put up the changes soon and ask for you to take a look to see if it's gone the direction you want.

I started off with e.g. `data: Data` as a way to simplify while working toward correct/test-able behavior.